File: lib/services/portfolio.js
import db from '../database.js';

/**
 * Portfolio Service
 * Handles all portfolio project-related database operations
 */

export class PortfolioService {
  constructor() {
    this.filename = 'portfolio.json';
  }

  /**
   * Get all projects
   */
  async getAll() {
    try {
      const projects = await db.readData(this.filename, []);
      return projects.sort((a, b) => 
        new Date(b.updatedAt || b.createdAt).getTime() - new Date(a.updatedAt || a.createdAt).getTime()
      );
    } catch (error) {
      console.error('Error getting all projects:', error);
      return [];
    }
  }

  /**
   * Get active projects only
   */
  async getActive() {
    try {
      const projects = await this.getAll();
      return projects.filter(project => project.active !== false);
    } catch (error) {
      console.error('Error getting active projects:', error);
      return [];
    }
  }

  /**
   * Get project by ID
   */
  async getById(id) {
    try {
      const projects = await this.getAll();
      return projects.find(project => project.id === id) || null;
    } catch (error) {
      console.error('Error getting project by ID:', error);
      return null;
    }
  }

  /**
   * Create new project
   */
  async create(projectData) {
    try {
      const projects = await this.getAll();
      
      const newProject = {
        id: db.generateId('project'),
        createdAt: db.getTimestamp(),
        updatedAt: db.getTimestamp(),
        active: true,
        ...projectData
      };

      projects.push(newProject);
      await db.writeData(this.filename, projects);
      
      console.log('✅ Project created successfully:', newProject.id);
      return newProject;
    } catch (error) {
      console.error('Error creating project:', error);
      throw new Error('Failed to create project');
    }
  }

  /**
   * Update project
   */
  async update(id, updates) {
    try {
      const projects = await this.getAll();
      const index = projects.findIndex(project => project.id === id);
      
      if (index === -1) {
        throw new Error('Project not found');
      }

      projects[index] = { 
        ...projects[index], 
        ...updates,
        updatedAt: db.getTimestamp()
      };

      await db.writeData(this.filename, projects);
      
      console.log('✅ Project updated successfully:', id);
      return projects[index];
    } catch (error) {
      console.error('Error updating project:', error);
      throw new Error('Failed to update project');
    }
  }

  /**
   * Delete project
   */
  async delete(id) {
    try {
      const projects = await this.getAll();
      const filteredProjects = projects.filter(project => project.id !== id);
      
      if (filteredProjects.length === projects.length) {
        throw new Error('Project not found');
      }

      await db.writeData(this.filename, filteredProjects);
      
      console.log('✅ Project deleted successfully:', id);
      return true;
    } catch (error) {
      console.error('Error deleting project:', error);
      throw new Error('Failed to delete project');
    }
  }

  /**
   * Search projects
   */
  async search(searchTerm) {
    try {
      const projects = await this.getAll();
      
      if (!searchTerm) return projects;

      const term = searchTerm.toLowerCase();
      return projects.filter(project =>
        project.title?.toLowerCase().includes(term) ||
        project.shortDesc?.toLowerCase().includes(term) ||
        project.longDesc?.toLowerCase().includes(term) ||
        project.useCases?.some(useCase => useCase.toLowerCase().includes(term))
      );
    } catch (error) {
      console.error('Error searching projects:', error);
      return [];
    }
  }
}

// Create and export a single instance
const portfolioService = new PortfolioService();
export default portfolioService;

File: app/services/portfolio.py
from typing import List, Dict, Any, Optional
from .database import db

class PortfolioService:
    def __init__(self):
        self.filename = 'portfolio.json'
    
    async def get_all(self) -> List[Dict[str, Any]]:
        """Get all portfolio projects"""
        try:
            projects = await db.read_data(self.filename, [])
            return projects
        except Exception as e:
            print(f"Error getting all portfolio projects: {e}")
            return []
    
    async def get_by_id(self, project_id: str) -> Optional[Dict[str, Any]]:
        """Get portfolio project by ID"""
        try:
            projects = await self.get_all()
            return next((proj for proj in projects if proj.get('id') == project_id), None)
        except Exception as e:
            print(f"Error getting portfolio project by ID: {e}")
            return None
    
    async def create(self, project_data: Dict[str, Any]) -> Dict[str, Any]:
        """Create new portfolio project"""
        try:
            projects = await self.get_all()
            
            new_project = {
                'id': db.generate_id('project'),
                'createdAt': db.get_timestamp(),
                **project_data
            }
            
            projects.append(new_project)
            await db.write_data(self.filename, projects)
            
            print(f"✅ Portfolio project created successfully: {new_project['id']}")
            return new_project
        except Exception as e:
            print(f"Error creating portfolio project: {e}")
            raise Exception("Failed to create portfolio project")
    
    async def update(self, project_id: str, updates: Dict[str, Any]) -> Dict[str, Any]:
        """Update portfolio project"""
        try:
            projects = await self.get_all()
            index = next((i for i, proj in enumerate(projects) if proj.get('id') == project_id), -1)
            
            if index == -1:
                raise Exception("Portfolio project not found")
            
            projects[index] = {
                **projects[index],
                **updates,
                'updatedAt': db.get_timestamp()
            }
            
            await db.write_data(self.filename, projects)
            print(f"✅ Portfolio project updated successfully: {project_id}")
            return projects[index]
        except Exception as e:
            print(f"Error updating portfolio project: {e}")
            raise Exception("Failed to update portfolio project")
    
    async def delete(self, project_id: str) -> bool:
        """Delete portfolio project"""
        try:
            projects = await self.get_all()
            filtered_projects = [proj for proj in projects if proj.get('id') != project_id]
            
            if len(filtered_projects) == len(projects):
                raise Exception("Portfolio project not found")
            
            await db.write_data(self.filename, filtered_projects)
            print(f"✅ Portfolio project deleted successfully: {project_id}")
            return True
        except Exception as e:
            print(f"Error deleting portfolio project: {e}")
            raise Exception("Failed to delete portfolio project")

# Global service instance
portfolio_service = PortfolioService()
